### ISSUE TITLE
Non-zero status code returned while running Softmax node

### ISSUE THREAD

**Issue by axel588** [microsoft/onnxruntime#15255]
### Describe the issue

Microsoft.ML.OnnxRuntime.OnnxRuntimeException : '[ErrorCode:RuntimeException] Non-zero status code returned while running Softmax node. Name:'/encoder/block.0/layer.0/SelfAttention/Softmax' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(2340)\onnxruntime.DLL!00007FFDFBA7E70A: (caller: 00007FFDFBA7DC5B) Exception(3) tid(c28c) 80070057 Param�tre incorrect.
'

I get this issue when using onnx runtime direct ml in C#, 
Everything works perfectly fine when falling back to cpu.

### To reproduce

Using C# onnx runtime with direct ml on line :
var encoderResults = transformer.Run(encoderFeedss);
with encoderFeeds :
{"tokens":[2662,1395,6051,1],"attention_mask":[1,1,1,1]}

### Urgency

Low

### Platform

Windows

### OS Version

Windows 11, latest

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.14.1

### ONNX Runtime API

C#

### Architecture

X64

### Execution Provider

DirectML

### Execution Provider Library Version

DirectML Onnx Runtime

**Comment by fdwr** [comment]
- Is the .onnx model available?
- Hmm, do you have the line in your code? Line 2340 has evidently changed since your version: https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/dml/DmlExecutionProvider/src/MLOperatorAuthorImpl.cpp#LL2340C2-L2340C6
- Do you see any more information if you enable the D3D debug layer? ![image](https://user-images.githubusercontent.com/1809166/228382907-d3e22af9-05f7-484d-ae22-79c53c8fccfb.png)

**Comment by axel588** [comment]
Thanks for your response @fdwr
I enabled D3D debug layer; still I get this message :
```
2023-03-29 19:43:20.5426113 [E:onnxruntime:, sequential_executor.cc:494 onnxruntime::ExecuteKernel] Non-zero status code returned while running Softmax node. Name:'/encoder/block.0/layer.0/SelfAttention/Softmax' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(2340)\onnxruntime.DLL!00007FF998EDE70A: (caller: 00007FF998EDDC5B) Exception(3) tid(8c34) 80070057 Paramfail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
      Microsoft.ML.OnnxRuntime.OnnxRuntimeException: [ErrorCode:RuntimeException] Non-zero status code returned while running Softmax node. Name:'/encoder/block.0/layer.0/SelfAttention/Softmax' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(2340)\onnxruntime.DLL!00007FF998EDE70A: (caller: 00007FF998EDDC5B) Exception(3) tid(8c34) 80070057 Param?tre incorrect.

         at Microsoft.ML.OnnxRuntime.NativeApiStatus.VerifySuccess(IntPtr nativeStatus)
         at Microsoft.ML.OnnxRuntime.InferenceSession.RunImpl(RunOptions options, IntPtr[] inputNames, IntPtr[] inputValues, IntPtr[] outputNames, DisposableList`1 cleanupList)
         at Microsoft.ML.OnnxRuntime.InferenceSession.Run(IReadOnlyCollection`1 inputs, IReadOnlyCollection`1 outputNames, RunOptions options)
         at Microsoft.ML.OnnxRuntime.InferenceSession.Run(IReadOnlyCollection`1 inputs, IReadOnlyCollection`1 outputNames)
         at Microsoft.ML.OnnxRuntime.InferenceSession.Run(IReadOnlyCollection`1 inputs)
         ```
The onnx model is **t5-base**

**Comment by fdwr** [comment]
>The onnx model is t5

It would be easiest to diagnose if we could reproduce this locally. Please provide a link to the specific version of the T5 .onnx file you are using (e.g. on Huggingface or wherever you obtained it from?). Thanks.

**Comment by axel588** [comment]
@fdwr
Hey yes sorry
https://huggingface.co/lewington/browser-based-imagen/tree/main
everything is in browser-based-imagen.zip, the two onnx the t5 and the model for image

**Comment by rytisss** [comment]
Ran into the same issue.  Interestingly, the same build works on the machine with AMD integrated and NVidia dedicated GPU.  Although it does not work on another device with Intel integrated GPU.

I do not know if the hardware configuration is related to this issue.

**Comment by axel578** [comment]
@rytisss Yes I use the integrated AMD Gpu, if you use an external nvidia gpu, it should work, but I encounter this issue using direct ml with amd integrated gpu

**Comment by s-redko** [comment]
Hi guys, we run into the same issue with `NVidia RTX A6000` using `onnx.runtime.1.14.1` when using `DirectML` execution provider.
`2023-04-18 14:42:16.2489888 [E:onnxruntime:, sequential_executor.cc:494 onnxruntime::ExecuteKernel] Non-zero status code returned while running Softmax node. Name:'Softmax_328' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(2340)\onnxruntime.DLL!00007FF9C4E1E70A: (caller: 00007FF9C4E1DC5B) Exception(3) tid(5824) 80070057 The parameter is incorrect.`

When we run the same model on `Intel HD Graphics 5500` using `onnx.runtime.1.13.1` with `DirectML` we get slightly different exception.
`2023-04-19 10:06:15.7477147 [E:onnxruntime:, sequential_executor.cc:368 onnxruntime::SequentialExecutor::Execute] Non-zero status code returned while running Slice node. Name:'Slice_208' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(1857)\onnxruntime.DLL!00007FFFE0B5E680: (caller: 00007FFFE0B5DB83) Exception(3) tid(6dc) 80070057 The parameter is incorrect`

With `CPU` and `CUDA` everything runs fine.

**Comment by axel588** [comment]
Up to now, I have found neither a solution or alternative, this is blocking

**Comment by fdwr** [comment]
> we run into the same issue with NVidia RTX A6000 using onnx.runtime.1.14.1

@s-redko: Do you have the input sizes used for these two tensors below? My teammate hasn't gotten to this bug yet, and I tried to repro it briefly on my NVidia card which runs without error, but then I'm probably not using the same sizes as you. 🤔

`int64 tokens[batch_size,sequence_length]`
`int64 attention_mask[batch_size,sequence_length]`

**Comment by fdwr** [comment]
> we run into the same issue with NVidia RTX A6000 using onnx.runtime.1.14.1

@s-redko: Do you have the sizes you used for these two input tensors below? My teammate hasn't gotten to this bug yet, and I tried to repro it briefly on my NVidia card which runs without error, but then I'm probably not using the same sizes as you 🤔 (using `batch_size:1` and `sequence_length:4` from @axel588's examples above executed fine, consistent with axel's experience, `{"tokens":[2662,1395,6051,1],"attention_mask":[1,1,1,1]}`). Thanks.

`int64 tokens[batch_size,sequence_length]`
`int64 attention_mask[batch_size,sequence_length]`

```
onnxruntime_perf_test.exe -I -e dml -r 1 -f batch_size:1 -f sequence_length:4 D:\ai-models\t5\t5-model.onnx
...
Total inference time cost: 0.0375492 s
Total inference requests: 1
```

**Comment by axel578** [comment]
The issue on my side is possible to reproduce with amd graphic card / interlnal graphic card like 680m. On my side the issue is mostly with amd, not nvidia

**Comment by s-redko** [comment]
> @s-redko: Do you have the sizes you used for these two input tensors below? 

@fdwr I might have misled you, sorry about that. The errors I have reported was not from running `t5-model.onnx`. They are from running on our own model that we have converted from `PyTorch` I should have specified it in my original message.

I've build `onnxruntime` and run `t5-model.onnx` since then using command line you've posted earlier.
> onnxruntime_perf_test.exe -I -e dml -r 1 -f batch_size:1 -f sequence_length:4 D:\ai-models\t5\t5-model.onnx

Which worked fine on both `integrated intel` and `nvidia` gpus. Furthermore I've run `onnxruntime_perf_test` with our model using `dml` and it also run fine.
Which makes me think that we have a bug in our code that uses onnxruntime API however I find it strange that same code works fine with `CPU` and `CUDA` providers but not `DML`.

Let me know if you need any other information from me. Unfortunately I'm not able to share our model though.

**Comment by JarradPosey** [comment]
I was having a similar issue with `DirectML` where I was getting this error:

> [ErrorCode:RuntimeException] Non-zero status code returned while running Softmax node. Name:'Softmax_329' Status Message: D:\a\_work\1\s\onnxruntime\core\providers\dml\DmlExecutionProvider\src\MLOperatorAuthorImpl.cpp(2340)\onnxruntime.DLL!00007FFFB07EE70A: (caller: 00007FFFB07EDC5B) Exception(3) tid(b39c) 80070057 The parameter is incorrect.

My issue appeared to be due to the `DirectML` dlls missing from the output directory. This seems to only occur in .NET Standard / Framework projects when the library `Microsoft.ML.OnnxRuntime.DirectML` is a transitive reference. Similar to https://github.com/NuGet/Home/issues/4488.

My solution was to add a direct dependency on the library to my parent project, i.e:
``` csproj
<PackageReference Include="Microsoft.ML.OnnxRuntime.DirectML" Version="1.14.1" />
```
Hope this helps someone✌️

**Comment by nums11** [comment]
Closing as resolved.

**Comment by Wingseter** [comment]
The issue has already been closed, but I want to explain why I encountered this problem and how I resolved it.

This issue is because DirectML.dll is installed by default in the System32 folder on Windows 10 and above.

In my Windows, the default version of DirectML was 1.8, while OnnxRuntime 1.16.1 was using DirectML version 1.12.1 by default.

I was able to reproduce the issue after downgrading DirectML to version 1.8.

Therefore, if you encounter this problem, you should check if the DLL used by your program matches the exact version you need.

I encountered this issue in Unity and resolved it by loading DirectML.dll at startup.

**Comment by antonvalletas** [comment]
Bless your heart. It would be so useful if a version check was just implemented and threw an informative error (versions don't match?) instead of random runtime errors on random layers/nodes if versions don't actually match.

**Comment by RichardGe** [comment]
Just add a comment to confirm I had this exact same issue because my project was using the DirectML from sytem32 instead of the DirectML used to compile ONNX runtime. Copy the correct DLL in the exe folder solved the issue.

**Comment by Jose17-ml** [comment]
Hi @Wingseter,

>>> In my Windows, the default version of DirectML was 1.8, while OnnxRuntime 1.16.1 was using DirectML version 1.12.1 by default.

How to check DirectML version in "C:\Windows\System32\directml.dll" and in "onnxruntime"? 

Currently I'm using 
onnxruntime          1.18.0

**Comment by fdwr** [comment]
> How to check DirectML version in "C:\Windows\System32\directml.dll" and in "onnxruntime"?

Right-click / Properties:

![image](https://github.com/microsoft/onnxruntime/assets/1809166/053d6782-a121-490f-858a-cd439fcf4be1)
